libp2p/wire.nim

```diff
@@ -20,20 +20,42 @@ else:
 
 const
   TRANSPMA* = mapOr(
-    mapAnd(IP, mapEq("udp")),
-    mapAnd(IP, mapEq("tcp")),
-    mapAnd(mapEq("unix"))
+    UDP,
+    TCP,
+    UNIX,
   )
 
   RTRANSPMA* = mapOr(
-    mapAnd(IP, mapEq("tcp")),
-    mapAnd(mapEq("unix"))
+    TCP,
+    UNIX
   )
 
+proc resolveDns(ma: MultiAddress, domain: Domain = Domain.AF_UNSPEC, prefix: string = ""): MaResult[TransportAddress] =
+  try:
+    var
+      dnsbuf: array[256, byte]
+      pbuf: array[2, byte]
+
+    if ma[0].`?`.protoArgument(dnsbuf).`?` == 0:
+      err("Invalid DNS argument")
+    else:
+      let dnsval = string.fromBytes(dnsbuf)
+      if ma[1].`?`.protoArgument(pbuf).`?` == 0:
```

Comment:
I like this a bit better than the `?` sandwich `(?(?ma[1]).protoArgument(pbuf))`
But tell me if you want me to use the sandwich

Reply:
no particular opinion since none of them are pretty, except that the `()` is used elsewhere, which is a good discriminator in these cases. The alternative would be to try and figure out a better chaining syntax in general, but that's a bit beyond the scope of this PR.